src/app/views/query-runner/query-input/AutoComplete.tsx

```diff
@@ -0,0 +1,344 @@
+import { connect } from 'react-redux';
+import { bindActionCreators, Dispatch } from 'redux';
+
+import { IAutoCompleteProps, IAutoCompleteState } from '../../../../types/auto-complete';
+import * as autoCompleteActionCreators from '../../../services/actions/autocomplete-action-creators';
+import { parseSampleUrl } from '../../../utils/sample-url-generation';
+import { queryInputStyles } from './QueryInput.styles';
+import { cleanUpSelectedSuggestion } from './util';
+
+class AutoComplete extends Component<IAutoCompleteProps, IAutoCompleteState> {
+  private autoCompleteRef: React.RefObject<ITextField>;
+
+  constructor(props: IAutoCompleteProps) {
+    super(props);
+
+    this.autoCompleteRef = React.createRef();
+
+    this.state = {
+      activeSuggestion: 0,
+      filteredSuggestions: [],
+      suggestions: [],
+      showSuggestions: false,
+      userInput: this.props.sampleQuery.sampleUrl,
+      compare: ''
+    };
+  }
+
+  public setFocus() {
+    if (this.autoCompleteRef && this.autoCompleteRef.current) {
+      this.autoCompleteRef.current.focus();
+    }
+  }
+
+  public onChange = (e: any) => {
+    const { suggestions, showSuggestions, userInput: previousUserInput, compare } = this.state;
+    const userInput = e.target.value;
+
+    this.props.contentChanged(userInput);
+
+    this.setState({
+      userInput
+    });
+
+    if (showSuggestions && suggestions.length) {
+      this.filterSuggestions(userInput, previousUserInput, compare, suggestions);
+    }
+    this.initialiseAutoComplete(userInput);
+  };
+
+  public onClick = (e: any) => {
+    const selected = e.currentTarget.innerText;
+    this.appendSuggestionToUrl(selected);
+  };
+
+  private initialiseAutoComplete = (url: string) => {
+    const lastCharacter = url.substring(url.length - 1);
+    switch (lastCharacter) {
+      case '/':
+        this.requestForAutocompleteOptions(url);
+        break;
+
+      case '?':
+        this.requestForAutocompleteOptions(url);
+        break;
+
+      case '=':
+
+        if (url.includes('?$')) {
+          this.getParameterEnums(url);
+        }
+
+        break;
+
+      case ',':
+        this.getParameterEnums(url);
+        break;
+
+      case '&':
+        this.getQueryParameters();
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  public onKeyDown = (e: any) => {
+    const { activeSuggestion, filteredSuggestions, showSuggestions } = this.state;
+
+    switch (e.keyCode) {
+      case KeyCodes.enter:
+        if (showSuggestions) {
+          const selected = filteredSuggestions[activeSuggestion];
+          this.appendSuggestionToUrl(selected);
+        } else {
+          this.props.runQuery();
+        }
+        break;
+
+      case KeyCodes.up:
+        if (activeSuggestion === 0) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion - 1 });
+        break;
+
+      case KeyCodes.down:
+        if (activeSuggestion - 1 === filteredSuggestions.length) {
```

Comment:
I think this should be the other way around in order not to go out of bounds:
```suggestion
        if (activeSuggestion === filteredSuggestions.length - 1) {
```